keps/sig-api-machinery/customresources-pruning-and-defaulting.md

```diff
@@ -0,0 +1,510 @@
+and the value validation (usually done in the validation step of the API server handler pipeline for native types). 
+
+For CRDs we cannot distinguish both. This was first noticed by @lavalamp in https://github.com/kubernetes/kubernetes/pull/64907#issuecomment-397015030.
+
+#### Example 6
+
+> Assume 
+>
+>```json
+>{
+>   "properties": {
+>       "a": {"type": "string", "pattern": "<some-long-regex-for-ips>", "format": "ip"},
+>       "b": {"properties": {"x": {}, "y": {}}}
+>   }
+>}
+>```
+>
+> The type and properties are usually called schema validation, while regex patterns and formats are about values. The latter would be tested in the validation phase, the former would be checked during decoding for native types in the JSON decoder.
+
+**Remark:** the line between type and format is blurry. Format is optional to be processed by tools, and the value is open ended. In Kube we have some formats (like `date`) which correspond to custom JSON unmarshallers in the native types. That format would be verified during decoding, although technically a date is just a string. So we might want to replicate that behaviour at some point.
+
+For pruning and defaulting we have to apply the OpenAPI validation schema inside of the decoder step (the left box inside apiextensions-apiserver in the figure above). This is considerably earlier than for native types. This has a number of implications:
+* We have to apply full OpenAPI value validation (e.g. regular expressions, propositional evaluation) during decoding.
+* Our generic registry applies validation after defaulting. We need the other way around for CRDs.
+* Our generic registry expects defaulting to always succeed (no error result type). CRD validation needed by defaulting can fail.
+
+To avoid these and to get sane semantics, **we propose to split the CRD validation into two steps:**
+1. During decoding we validate using a skeleton schema which lacks value validations and propositional logic (`anyOf`, `allOf`, `oneOf`, `not`).
+2. During defaulting we validate using the skeleton schema and apply the resulting defaulter.
+During standard generic registry validation phase we validate using the full validation schema.
+
+Because the first step succeeds, also the second step succeeds. So we don’t need error handling during defaulting. The third step will catch wrong types of the used defaults.
+
+Note that defaulting – like for native types – will be done
+1. before conversion of the incoming API objects and of objects read from storage and
+2. after mutating admission webhooks (compare [k8s.io/apiserver/pkg/admission/plugin/webhook/mutating/dispatcher.go](https://github.com/kubernetes/kubernetes/blob/e4c219df42c77ecb8f0588197072bef81bca7429/staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/mutating/dispatcher.go#L152)).
+
+## Formal Proposal – following pruning option 1
```

Comment:
I think I would rather approach it this way: reject bad schemas early.

1. Define a schema that expresses exactly and only the things we want to allow folks to express (i.e., things that can be done with go structs).
2. Translate both ways between that schema and OpenAPI spec.
3. Round-trip a CRD's `validation` tag through that. If it comes through unchanged, it is acceptable. Otherwise, it is unacceptable.

How exactly to roll this out into the validation of CRDs given that it's tightening our restrictions is another discussion.

I appreciate the cleverness that went into this section of the KEP. I think, however, we should keep in mind the overall goal of this chunk of the code base: we're allowing more ways of getting a specific chunk of functionality. OpenAPI spec allows a bunch of features that are about different functionality. Since we're not trying to implement arbitrary APIs, we shouldn't be clever in supporting those features, we should be preventing people from using them.

Nothing made by kube-openapi or kube-builder should be using these features. We are not taking features away from people by rejecting specs mixing schema and value validations.